titan-compiler/coder.lua

```diff
@@ -271,273 +300,385 @@ local function set_heap_slot(typ, dst, src, parent)
+            ${RET_DECL} = ${TITAN_CALL};
+            ${SET_SLOT}
+            api_incr_top(L);
+        ]], {
+            TITAN_CALL = titan_call,
+            RET_DECL = c_declaration(ret),
+            SET_SLOT = set_stack_slot(ret_typ, "s2v(L->top)", ret.name),
+        })
+    else
+        error("not implemented")
+    end
+
+    return util.render([[
+        static int ${LUA_ENTRY_POINT}(lua_State *L)
+        {
+            ${SET_BASE}
+            ${CHECK_NARGS}
+            ${CHECK_TYPES}
+            ${GET_ARGS}
+            ${SET_RETURN}
+            return ${NRET};
+        }
+    ]], {
+        LUA_ENTRY_POINT = tl_node._lua_entry_point,
+        SET_BASE = set_base,
+        CHECK_NARGS = check_nargs,
+        CHECK_TYPES = table.concat(check_types, "\n"),
+        GET_ARGS = table.concat(get_args, "\n"),
+        SET_RETURN = set_return,
+        NRET = c_integer(#tl_node._type.rettypes),
+    })
+end
+
+local function generate_luaopen_upvalue_array(prog, ctx)
+    local parts = {}
+
+    for _, tl_node in ipairs(prog) do
+        if tl_node._global_index then
+            local arr_slot = util.render([[ &${ARR}[${I}] ]], {
+                ARR = ctx.upvalues.array.name,
+                I = c_integer(tl_node._global_index),
+            })
+
+            table.insert(parts,
+                string.format("/* %s */", ast.toplevel_name(tl_node)))
+
+            local tag = tl_node._tag
+            if     tag == ast.Toplevel.Func then
+                ctx:begin_scope()
+                local func    = ctx:new_cvar("CClosure*")
+                local func_cl = "((Closure*)" .. func.name .. ")"
+                table.insert(parts,
                     util.render([[
-                        static int ${LUA_ENTRY_POINT}(lua_State *L)
-                        {
-                            ${ARGS_DECL}
-                            ${RET_INIT} ${TITAN_ENTRY_POINT}(${ARGS});
-                            ${SET_RET}
-                            return ${NRET};
-                        }
-                    ]], {
+                        ${FUNC_DECL} = luaF_newCclosure(L, 1);
+                        ${FUNC}->f = ${LUA_ENTRY_POINT};
+                        sethvalue(L, &${FUNC}->upvalue[0], ${UPVALUES});
+                        ${SET_SLOT}
+                    ]],{
                         LUA_ENTRY_POINT = tl_node._lua_entry_point,
-                        TITAN_ENTRY_POINT = tl_node._titan_entry_point,
-                        RET_INIT = ret_init,
-                        ARGS_DECL = table.concat(args_decl, "\n"),
-                        ARGS = table.concat(args, ", "),
-                        SET_RET = set_ret,
-                        NRET = c_integer(#tl_node._type.rettypes)
+                        UPVALUES = ctx.upvalues.table.name,
+                        FUNC = func.name,
+                        FUNC_DECL = c_declaration(func),
+                        SET_SLOT = set_heap_slot( -- TODO
```

Comment:
TODO what?

Reply:
I honestly don't remember... But this TODO is gone in the current versioni of the code. I must have fixed whatever it was.